### DISCUSSION TITLE
.NET 1.0.1 Unit test strategy

### DISCUSSION THREAD

**Discussion by jimmy-vo-librestream** [microsoft/semantic-kernel#4575] [Q&A] [answered]
I am working on upgrading 1.0.0-beta8 to 1.0.1, and I found it is not possible to mock Kenel, Function, or Pluggin. For example I have an extension function that my business logic calling to:
```
    public static async Task<string> InvokePromptAsync(
        this Kernel kernel,
        KernelArguments variables,
        string template,
        OpenAIPromptExecutionSettings requestSettings,
        CancellationToken cancellationToken = default)
    {
        KernelFunction skFunction = kernel.CreateFunctionFromPrompt(template, requestSettings);
        var result = await kernel.InvokeAsync<string>(skFunction, variables, cancellationToken);
        return result!;
    }
```
In 1.0.0-beta8, I was able to mock `ISKFunction`, `IReadOnlyFunctionCollection`, and `IKernel`. But they are removed in 1.0.1. `KernelFunction` is now an abstract class with internal constructor, so there is no way I can create a `MockKernelFunction` that wraps KernelFunction and overrides `ValueTask<FunctionResult> InvokeCoreAsync(Kernel kernel, KernelArguments arguments, CancellationToken cancellationToken)`.

I am currious how you guys do it?

**Comment by pratibw** [comment]
Does anyone know how to setup the mock for InvokeAsync?

this.kernelMock.Setup(k => k.InvokeAsync(It.IsAny<KernelFunction>(), It.IsAny<KernelArguments>())).ReturnsAsync(It.IsAny<FunctionResult>());
            
 this is throwing an error and I am not able to mock the call.

**Comment by madsbolaris** [comment, marked as answer]
Per @markwallace-microsoft guidance on #4589, you can take a look at our existing unit tests here for `KernelFunction`:

[KernelFunctionFromMethodTests1.cs](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/SemanticKernel.UnitTests/Functions/KernelFunctionFromMethodTests1.cs)
[KernelFunctionFromMethodTests2.cs](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/SemanticKernel.UnitTests/Functions/KernelFunctionFromMethodTests2.cs)
[KernelFunctionFromPromptTests.cs](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/SemanticKernel.UnitTests/Functions/KernelFunctionFromPromptTests.cs)

For your use case you can mock the AI service and add that to the Kernel you use in your unit tests. We're going to take a look at providing guidance around unit testing.

**Comment by dmytrostruk** [comment]
@jimmy-vo-librestream @pratibw Here is a blog post about unit testing of Semantic Kernel components:
https://devblogs.microsoft.com/semantic-kernel/unit-testing-with-semantic-kernel/

Hope it will help!

**Reply by vikramhn** [reply]
great blog on unit testing of Semantic Kernel components, any possibility of putting that in a repo with a sample and those unit tests so people can see it in action
